### PR TITLE
[R4R] use priority lock for blockchain

### PR DESCRIPTION
**Pull request by forcodedancing**
### Description

This pr tries to improve block importing by using priority lock for `blockchain`, which will give block insertion higher priority than others.

### Rationale

During block insertion there are other subroutines also try to access the ongoing blockchain, which could slowdown the main routine of block insertion.

### Result
* Before
```
Dropped 527 nodes (cum <= 0.01mins)
Showing top 40 nodes out of 126
      flat  flat%   sum%        cum   cum%
  1.99mins 87.30% 87.30%   1.99mins 87.30%  sync.(*Mutex).Unlock (partial-inline)
  0.27mins 11.94% 99.24%   1.91mins 83.79%  sync.(*RWMutex).Unlock
  0.02mins  0.76%   100%   0.02mins  0.76%  sync.(*RWMutex).RUnlock (inline)
         0     0%   100%   0.02mins  0.83%  github.com/VictoriaMetrics/fastcache.(*Cache).Set
         0     0%   100%   0.02mins  0.83%  github.com/VictoriaMetrics/fastcache.(*bucket).Set
         0     0%   100%   0.77mins 33.82%  github.com/ethereum/go-ethereum/core.(*BlockChain).InsertChain
         0     0%   100%   0.02mins  0.92%  github.com/ethereum/go-ethereum/core.(*BlockChain).insertChain
         0     0%   100%   0.01mins  0.65%  github.com/ethereum/go-ethereum/core.(*BlockChain).writeBlockWithState
         0     0%   100%   0.01mins  0.65%  github.com/ethereum/go-ethereum/core.(*BlockChain).writeHeadBlock
```

* After
```
Dropped 525 nodes (cum <= 0.01mins)
Showing top 40 nodes out of 125
      flat  flat%   sum%        cum   cum%
  1.35mins 59.00% 59.00%   1.35mins 59.00%  sync.(*Mutex).Unlock (partial-inline)
  0.92mins 40.09% 99.09%   1.95mins 85.46%  sync.(*RWMutex).Unlock
  0.02mins  0.91%   100%   0.02mins  0.91%  sync.(*RWMutex).RUnlock (inline)
         0     0%   100%   0.02mins  0.77%  github.com/VictoriaMetrics/fastcache.(*Cache).Set
         0     0%   100%   0.02mins  0.77%  github.com/VictoriaMetrics/fastcache.(*bucket).Set
         0     0%   100%   0.02mins  1.05%  github.com/ethereum/go-ethereum/common/prlock.(*Prlock).LockHigh
         0     0%   100%   0.75mins 32.90%  github.com/ethereum/go-ethereum/common/prlock.(*Prlock).UnlockHigh (inline)
         0     0%   100%   0.80mins 35.07%  github.com/ethereum/go-ethereum/core.(*BlockChain).InsertChain
         0     0%   100%   0.03mins  1.13%  github.com/ethereum/go-ethereum/core.(*BlockChain).insertChain
         0     0%   100%   0.02mins  0.84%  github.com/ethereum/go-ethereum/core.(*BlockChain).writeBlockWithState
         0     0%   100%   0.02mins  0.84%  github.com/ethereum/go-ethereum/core.(*BlockChain).writeHeadBlock
```

We can see for using priority lock, `InsertChain` block time increased, from `0.77` to `0.8`. These two nodes did not response to any other requests/workload. There is a chance that priority can work better when there are other workloads. 
`AccountsIntermediateRoot` block time has beed decreased.

### Changes

Notable changes: 
* add priority lock for block insertion

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -1669,8 +1673,8 @@ func (bc *BlockChain) writeKnownBlock(block *types.Block) error {
 
 // WriteBlockWithState writes the block and all associated state to the database.
 func (bc *BlockChain) WriteBlockWithState(block *types.Block, receipts []*types.Receipt, logs []*types.Log, state *state.StateDB, emitHeadEvent bool) (status WriteStatus, err error) {
-	bc.chainmu.Lock()
-	defer bc.chainmu.Unlock()
+	bc.chainLock.LockLow()
```
Comment (guagualvcha) [inline core/blockchain.go:1676]: Why use `LockLow` here? I think `WriteBlockWithState` is on the critical path.

File: core/state/state_object.go [hunk 2]
```diff
@@ -367,7 +367,7 @@ func (s *StateObject) updateTrie(db Database) Trie {
 		}
 		// If state snapshotting is active, cache the data til commit
 		if s.db.snap != nil {
-			s.db.snapMux.Lock()
+			s.db.snapStorageMux.Lock()
```
Comment (guagualvcha) [inline core/state/state_object.go:370]: So are we using the false lock before?

File: core/blockchain.go (hunk 1 above)
Comment (forcodedancing) [reply core/blockchain.go:1676]: yes, you are right

File: core/blockchain.go [hunk 3]
```diff
@@ -591,15 +593,17 @@ func (bc *BlockChain) SetHead(head uint64) error {
 }
 
 func (bc *BlockChain) tryRewindBadBlocks() {
-	bc.chainmu.Lock()
-	defer bc.chainmu.Unlock()
```
Comment (setunapo) [inline core/blockchain.go:595]: lock scope has been changed. but it should be ok.

Comment (forcodedancing) [reply core/blockchain.go:595]: yes, a bit of refactor on purpose

File: common/prlock/prlock.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
+// Package prlock implements a priority lock data structure.
+package prlock
+
+import "sync"
+
+// Prlock implements a priority lock data structure.
+type Prlock struct {
+	data sync.RWMutex // Data mutex
+	next sync.Mutex   // Next to access mutex
```
[the remark this answers is not in the document]
Comment (forcodedancing) [reply common/prlock/prlock.go:9]: that's my understanding, i cannot find any detail explain on this.

[the remark this answers is not in the document]
Comment (setunapo) [reply common/prlock/prlock.go:9]: ok, it make sense, it make the highPrLock with greater priority.
I thought Three Locks could provide: `highPrLock`, `medieumPrLock` and `lowPrLock`

**Comment by xiyuewang3** [comment]
Great changes

**Pull request closed without merging**